imageio/core/imopen.py

```diff
@@ -180,6 +203,34 @@ def loader(request, **kwargs):
         request.finish()
         raise err_type(err_msg) from err_from
 
+    # fast-path based on format_hint
+    if request.format_hint is not None:
+        for candidate_format in known_extensions[format_hint]:
+            for plugin_name in candidate_format.priority:
+                config = known_plugins[plugin_name]
+
+                # v2 compatibility; delete in v3
+                if legacy_mode and not config.is_legacy:
+                    continue
+
+                try:
+                    candidate_plugin = config.plugin_class
+                except ImportError:
+                    # not installed
+                    continue
+
+                try:
+                    plugin_instance = candidate_plugin(request, **kwargs)
+                except InitializationError:
+                    # file extension doesn't match file type
+                    continue
+
+                return plugin_instance
+        else:
+            raise IOError(
+                f"The ImageResource can not be opened as a `{format_hint}` file."
+            )
```

Comment:
Actually, as it is now, the PR would only check plugins that are known to read the hinted format and fail otherwise. We can remove this part (or make it a warning) which would fall back to checking all plugins if the expected plugins don't work out.

@almarklein Which of these two approaches do you think is cleaner?

Reply:
I'm not sure what you mean, but I'd expect this:
* if plugin is given, that is the only plugin that will be used, and will fail otherwise.
* if format-hint is given, it will try to find a plugin that satisfies that format, and if it fails, will try harder, with a warning,
* if both are given, only the selected plugin is used, and again the format is a hint.

But why should the format be a hint? What about:
* if plugin is given, that is the only plugin that will be used, and will fail otherwise.
* if format is given, it will use a plugin that can read that format, and it will be read as that format, or it will fail.
* if both are given, it *has* to be the given format *and* plugin.

Reply:
> But why should the format be a hint? [...] and it will be read as that format, or it will fail.

It is currently a hint because we would struggle to enforce/guarantee that the ImageResource is indeed read as the desired format. 

From an interface design standpoint we can't control (or assert) if a backend actually opens the file using the desired format. Worse, some backends may not support enforcing a format, and then the plugin writer is stuck between a rock and a hard place, because the ImageIO API requires that you can force a format, but the backend doesn't permit this easily. The `plugin` argument on the other hand, we can (and do) enforce ourselves, so for that one we can give the user a guarantee that it will be that plugin or nothing.

From a practical perspective, you might also not get too much out of enforcing that a ImageResource is treated as a certain format. E.g., just because you enforce that the resource is treated as TIFF doesn't tell you anything about the data you will get (colorspace, shape, ndims, planar/interleaved aka. channel_axis). You'd have similar problems with videos (almost any container/format supports multiple codecs and pixel formats/colorspaces) and PNG (supports multiple colorspaces). Not sure about other formats. What I'm trying to say is that I'm not entirely sure what a user may gain from enforcing a format.

Further, if you do have a file with the wrong extension (and you don't know which one it should be/ what's inside the file) then ImageIO will already take care of that during fallback. It will first try the plugins associated with that extension and, if that fails, it will do a broader search considering all available plugins (which will read almost anything under the sun at this point).

> if format-hint is given, it will try to find a plugin that satisfies that format, and if it fails, will try harder, with a warning,

Okay, then I will change it to a warning (the other points are as you listed in what you'd expect). 

We can discuss the utility and desired behavior of enforcing a format further, and - based on the outcome - I will make further adjustments as needed.

Reply:
Ok, lets move on with that direction then.